File: E_n_meb/Source/RunCase_6_lemma2.m
#ifdef MAIN_LEMMA_2

#import "PathAlg.alg.h"
#import "KoefMatrix.h"
#import "PrintUtils.h"
#import "Utils.h"

void CreateMatrix2(KoefIntMatrix *matrix);
int RankLemma2(KoefIntMatrix *matrix);

//---------------------------------------------------------------------------------
BOOL _RunCase() {
    NSInteger n = PathAlg.alg.n;
    NSInteger s = PathAlg.alg.s;

    WriteLog(0, "N=%d, S=%d, char=%d",  n, s, PathAlg.alg.charK);

    for (NSInteger i = 0; i < 2 * PathAlg.alg.twistPeriod; i++) {
        KoefIntMatrix *matrix = [[KoefIntMatrix alloc] initWithSize:s];
        CreateMatrix2(matrix);

        NSInteger rk2 = RankLemma2(matrix);

        NSInteger rk1 = [matrix rank];

        if (rk2 && rk1 != rk2) {
            WriteLog(1, "%d (must be %d)! N=%d, S=%d, char=%d, matrix:", rk2, rk1, n, s, [PathAlg.alg charK]);
            printKoefIntMatrix(matrix, 0, 0);
        }

        [matrix release];
    }
    return NO;
}

//---------------------------------------------------------------------------------
void CreateMatrix2(KoefIntMatrix *matrix) {
    NSInteger s = PathAlg.alg.s;

    for (NSInteger i = 0; i < s; i++) {
        NumInt* n = matrix.rows[i][i];
        [n setIntValue:(rand() % 2) ? 1 : -1];
        n = matrix.rows[[PathAlg myModS:i + 1)][i];
        [n setIntValue:(rand() % 2) ? 1 : -1];
    }
}

//---------------------------------------------------------------------------------
int RankLemma2(KoefIntMatrix *matrix) {
    NSInteger charK = PathAlg.alg.charK;
    NSInteger n = (NSInteger)matrix.rows.count;

    if (charK == 2)
        return n - 1;

    NSInteger k = 1;
    for (NSInteger i = 0; i < n; i++) {
        NumInt* n1 = matrix.rows[i][i];
        k *= n1.intValue;
        n1 = matrix.rows[[PathAlg myMod:i + 1 mod: n]][i];
        k *= n1.intValue;
    }

    return (k == minusDeg(n)) ? n - 1 : n;
}

#endif

File: E_n_meb/Source/RunCase_6_lemma2.swift
//
//  Created by M on 02.04.17.
//
//

import Foundation

struct RunCase_6_lemma2
{
    static func runCase() -> Bool
    {
        let s = PathAlg.alg.s
        OutputFile.writeLog(2, "N=%d, S=%d, Char=%d",  PathAlg.alg.n, PathAlg.alg.s, PathAlg.alg.charK)
        for _ in 0...2 * PathAlg.alg.twistPeriod {
            let matrix = KoefIntMatrix(size: s)!
            createMatrix2(matrix)

            let rk2 = rankLemma2(matrix)
            let rk1 = matrix.rank()

            if rk2 != 0 && rk1 != rk2 {
                OutputFile.writeLog(1, "%d (must be %d)! N=%d, S=%d, char=%d, matrix:",
                                    rk1, rk2, PathAlg.alg.n, s, PathAlg.alg.charK)
                printKoefIntMatrix(matrix, 0, 0)
                return true
            }
        }
        return false
    }

    private static func createMatrix2(_ matrix: KoefIntMatrix)
    {
        let s = PathAlg.alg.s
        for i in 0 ..< s {
            matrix.rows()[i][i].intValue = arc4random_uniform(2) == 0 ? 1 : -1
            matrix.rows()[myModS(i + 1)][i].intValue = arc4random_uniform(2) == 0 ? 1 : -1
        }
    }

    private static func rankLemma2(_ matrix: KoefIntMatrix) -> Int
    {
        let charK = PathAlg.alg.charK
        let n = matrix.rows().count

        if charK == 2 {
            return n - 1
        }

        var k = 1
        for i in 0 ..< n {
            k *= matrix.rows()[i][i].intValue
            k *= matrix.rows()[myMod(i + 1, mod: n)][i].intValue
        }

        return (k == minusDeg(n)) ? n - 1 : n;
    }
}
